### PR TITLE
PR: feat/08-autocomplete-customer-profiles-invoiceform

### LINKED ISSUE #50: Phase [8] Autocomplete Kunden & Profile
---
name: Feature
about: Neues Feature (UI/BE/Build)
labels: ["feature"]
---

## Kontext
Einbindung von Autocomplete-Feldern für Kunde und Profil.  
Backend liefert Daten aus `/customers` und `/profiles`.

## Ziel / Scope
- GET-Requests an Backend
- Combobox (shadcn/ui) für Kunden und Profile
- Auswahl speichert ID im Form-Value

## Akzeptanzkriterien / DoD
- [ ] Combobox funktioniert für beide Felder
- [ ] API-Fetch korrekt
- [ ] Auswahl speichert ID
- [ ] Kein Filter im FE, nur BE-Response

## Tasks
- [ ] API Service Layer erstellen (fetch wrapper)
- [ ] GET /customers + /profiles anbinden
- [ ] shadcn Combobox-Komponente konfigurieren
- [ ] Integration in Invoice Form

## Risiken / Out-of-Scope
- Keine Suche im Frontend
- Kein Caching der Daten

## Referenzen (Docs)
- shadcn/ui Combobox: https://ui.shadcn.com/docs/components/combobox  
- FastAPI Endpoints: `/customers`, `/profiles`

### PR DESCRIPTION
**Pull request by wontknow**
## Zusammenfassung
Implementierung von Autocomplete-Combobox-Feldern für Kunden und Profile in der Invoice Create Form.
Diese PR ersetzt die bisherigen Text-Input- und Select-Felder durch:
- **Kunden-Suchfeld**: Debounced Search mit Inline-Dropdown (300ms, min 2 chars)
- **Profil-Combobox**: Select mit ID-basierter Speicherung
- **Auto-Create Flow**: Wenn Kunde nicht gefunden, wird bei Submit automatisch erstellt
- **Form Schema**: Umbenennung von Namen zu IDs (bessere API-Integration)
- **Service Layer**: Search-Methoden für beide Entitäten

## Verwandte Issues
Closes #50 

#### `src/types/invoice-form.ts`
- **Old**: `customer_name: string`, `profile_name: string`
- **New**: `customer_id: number | null`, `profile_id: number`
- Rationale: IDs erlauben direkte API-Nutzung ohne Name-Lookup, verhindert Name-Konflikte
- `customer_id: null` = Trigger für Auto-Create on submit

#### `src/services/profiles.ts`
- ✅ `ProfilesService.search(query: string)` → `GET /profiles/search?q=...`
  - Min 2 chars validation
  - Logging: 🔍 start, ✅ results, ❌ error
  - Returns filtered Profile[]

#### `src/services/customers.ts`
- ✅ `CustomersService.search(query: string)` → `GET /customers/search?q=...`
  - Min 2 chars validation
  - Logging: 🔍 start, ✅ results, ❌ error
  - Returns filtered Customer[]

#### `src/features/invoices/InvoiceForm.tsx`

**State Management**:
- Added: `customerSearchResults` (Dropdown results)
- Added: `isSearchingCustomers` (Loading state)
- Added: `customerSearchInput` (Search text, separate from form value)
- Purpose: Separate search input from form ID storage for auto-create flow

**Kunden-Feld (NEW)**:
```tsx
<Input 
  placeholder="Kundennamen eingeben..."
  value={customerSearchInput}
  onChange={(e) => {...}} // Debounced search on every keystroke
/>
// Inline Dropdown mit clickable Results
{customerSearchResults.map(customer => (
  <button onClick={() => field.onChange(customer.id)}>
    {customer.name}
  </button>
))}
// Auto-create Indicator
{customerSearchInput && field.value === null && (
  <FormDescription>Neuer Kunde wird automatisch erstellt</FormDescription>
)}
```

**Profil-Feld (UPDATED)**:
- Changed: Select mit `value={field.value}` (string) → `value={String(profile_id)}`
- Changed: `onValueChange={(v) => field.onChange(v)}` → `onValueChange={(v) => field.onChange(parseInt(v))}`
- Storage: ID nicht Name

**Form Submission (Updated onSubmit)**:
```typescript
// Step 1: Resolve customer_id
if (data.customer_id !== null) {
  // Use existing customer
  customer_id = data.customer_id;
} else {
  // Auto-create: POST /customers with customerSearchInput name
  const newCustomer = await CustomersService.create({
    name: customerSearchInput.trim(),
    address: null,
    city: null,
  });
  customer_id = newCustomer.id;
}

// Step 2: Build payload with IDs (not names)
const payload = {
  date: data.date,
  customer_id,      // number
  profile_id: data.profile_id, // number
  total_amount: calculatedTotal,
  invoice_items: data.invoice_items,
  include_tax: data.include_tax,
  tax_rate: data.tax_rate,
  is_gross_amount: data.is_gross_amount,
};

// Step 3-5: Create, fetch, success
```

**Logging (Enhanced)**:
- 📄 = Flow start
- 🔍 = Search operations
- ✅ = Success operations
- ❌ = Errors
- 📤 = Outgoing payloads
- 📥 = Incoming responses
- ⚠️ = Warnings (auto-create)

#### `src/features/invoices/InvoiceForm.test.tsx`
- ✅ Updated selector: `getByLabelText(/Kundenname/)` → `getByPlaceholderText(/Kundennamen eingeben/)`
- ✅ Updated selector: `getByLabelText(/Profil/)` → `getByLabelText(/Profil/i)`
- ✅ All 51 tests passing
- Tests cover:
  - Rendering all fields ✓
  - Invoice number preview ✓
  - Profile/Customer fields existence ✓
  - Tax field visibility ✓
  - Item management (add/remove/limit) ✓
  - Total calculations ✓
  - Form validation ✓
  - Responsive design ✓

### Backend - No Changes
- Routes `/profiles/search` und `/customers/search` bereits vorhanden
- Backend liefert bereits gefilterte Results (ilike mit Escaping)

### Setup
```bash
# Terminal 1: Backend
cd backend
python -m uvicorn main:app --reload --port 8000

# Terminal 2: Frontend
cd frontend
pnpm run dev
# Öffne http://localhost:3000/invoices/create
```

### Happy Path Test

1. **Form Load**:
   - ✓ Rechnungsnummer Preview zeigt sich (z.B. "25 | 001")
   - ✓ Profil-Dropdown ist leer (bereit zum Laden)

2. **Profile Selection**:
   - Klick auf Profil-Select → Dropdown öffnet
   - Wähle Profil (z.B. "Hauptprofil")
   - ✓ Value wird gespeichert (ID, nicht Name)

3. **Customer Search** (NEW):
   - Klick auf Customer-Input
   - Tippe "test" → Wartet 300ms
   - ✓ Suche startet: `🔍 Searching customers: test`
   - ✓ Results erscheinen: Dropdown mit gefundenen Kunden
   - Klick auf Kunde → Selected, Input wird mit Name gefüllt
   - ✓ `field.value` speichert die ID

4. **Customer Auto-Create** (NEW):
   - Tippe Customer-Name: "Neue Testfirma GmbH"
   - Beende Search (leer = kein Match)
   - ✓ Text "Neuer Kunde wird automatisch erstellt" erscheint
   - Kein Auswahl-Dropdown

5. **Add Items & Submit**:
   - Füge Items hinzu (Beschreibung, Menge, Preis)
   - Aktiviere "Rechnung versteuern" + Steuersatz
   - Klick "Rechnung erstellen"

6. **Browser Console (Frontend Logging)**:
   ```
   📄 Starting invoice creation flow...
   Form data: {profile_id: 1, customer_id: null, items: [...]}
   🔍 Customer not selected, creating new with name: Neue Testfirma GmbH
   ✅ Customer created: {id: 456, name: "Neue Testfirma GmbH", ...}
   📤 Sending invoice payload: {customer_id: 456, profile_id: 1, ...}
   ✅ Invoice created successfully: {id: 999, number: "25 | 002", ...}
   📥 Fetched created invoice: {id: 999, number: "25 | 002", ...}
   ```

7. **Alert & Reset**:
   - ✓ Alert: "✅ Rechnung erfolgreich erstellt! Rechnungsnummer: 25 | 002"
   - ✓ Form resets
   - ✓ Invoice number preview aktualisiert → "25 | 003"

### Edge Cases

1. **Existing Customer**:
   - Tippe bestehenden Kundennamen
   - Ergebnisse erscheinen
   - Wähle einen aus
   - ✓ `customer_id` wird mit ID gefüllt (nicht null)
   - Submit → verwendet diese ID, keine Auto-Create

2. **Search Min Length**:
   - Tippe einzelnes Character "a"
   - ✓ Keine Search (min 2 chars)
   - Tippe zweites Character "ab"
   - ✓ Suche startet

3. **Empty Customer Name for Auto-Create**:
   - Nicht eingeben, direkt Submit
   - ✓ Error: "Bitte geben Sie einen Kundennamen ein"

4. **Validation**:
   - Leeres Formular Submit
   - ✓ Schema validates: `profile_id` is required, items min 1
   - ✓ No API call

## Browser Console Output Beispiel

```
✅ InvoiceForm mounted
📄 Starting invoice creation flow...
Form data: {
  profile_id: 1,
  customer_id: null,
  date: "2025-11-10",
  is_gross_amount: false,
  include_tax: true,
  tax_rate: 0.19,
  invoice_items: [
    {description: "Beratung", quantity: 5, price: 100},
    {description: "Entwicklung", quantity: 10, price: 80}
  ]
}
⚠️ Customer not selected, creating new with name: Brand New Inc.
📤 POST /customers/ payload: {name: "Brand New Inc.", address: null, city: null}
✅ Customer created: {id: 789, name: "Brand New Inc.", ...}
📤 POST /invoices/ payload: {
  date: "2025-11-10",
  customer_id: 789,
  profile_id: 1,
  total_amount: 1300.00,
  invoice_items: [...],
  include_tax: true,
  tax_rate: 0.19,
  is_gross_amount: false
}
✅ Invoice created successfully: {id: 2000, number: "25 | 010", ...}
📥 GET /invoices/2000
✅ Fetched created invoice: {id: 2000, number: "25 | 010", ...}
✅ Rechnung erfolgreich erstellt! Rechnungsnummer: 25 | 010
```

## Checkliste (DoD)

- [x] Funktionalität erfüllt (Akzeptanzkriterien)
  - [x] Profil-Combobox funktioniert
  - [x] Customer-Search mit Debounce (300ms)
  - [x] Auswahl speichert ID (nicht Name)
  - [x] Customer Auto-Create Flow
  - [x] Kein Frontend-Filter (nur Backend-Response)

- [x] Readme/Docs aktualisiert (falls nötig)
  - [x] Diese PR Dokumentation erstellt
  - [x] Test Flow dokumentiert
  - [x] Browser Console Logging beispiel

- [x] Keine Secrets/Keys committed
  - [x] .env nicht committed
  - [x] Nur Code & Config

- [x] Lint/Build grün
  - [x] pnpm format ✓
  - [x] ESLint ✓ (0 errors)
  - [x] TypeScript ✓ (0 errors)

- [x] Manuelle Tests (happy path & edge cases)
  - [x] Profile selection & ID storage
  - [x] Customer search with debounce
  - [x] Search results dropdown
  - [x] Auto-create flow
  - [x] Form validation
  - [x] Submit with IDs

- [x] CORS geprüft (Dev/Prod)
  - [x] ALLOWED_ORIGINS already configured in previous PR
  - [x] localhost:3000 ✓
  - [x] 192.168.2.116:3000 (LAN) ✓
  - [x] tauri://localhost ✓

- [ ] PDF erzeugt & visuell geprüft
  - Not in scope (Invoice Creation first, PDF later)

- [ ] (optional) E-Rechnung Validierung dokumentiert
  - Not in scope (Core invoice structure first)

## Breaking Changes

⚠️ **Form Schema Breaking Change**:
- Old: `{customer_name: string, profile_name: string, ...}`
- New: `{customer_id: number | null, profile_id: number, ...}`
- **Impact**: Any code expecting form data structure needs update
- **Migration**: Update form consumers to use IDs instead of names

### Why Separate customerSearchInput State?
Form stores `customer_id: number | null` (for API).
But for auto-create, we need the typed name (string) from the search input.
Solution: Keep `customerSearchInput` in component state, separate from form.

### Why IDs in Form?
- Direct API compatibility (backend expects customer_id, profile_id)
- Prevents name conflicts (multiple customers with same name)
- Simpler validation (just check non-null numbers)
- Better performance (no name-based lookups on submit)

### Search Implementation
- Frontend: No filtering logic. Just pass query to backend.
- Backend: ilike with proper escaping (already implemented)
- Debounce: 300ms standard for typeahead
- Min chars: 2 (backend requirement)

## Summary für Review
- **Scope**: Autocomplete-Integration für Kunden & Profile
- **Tests**: 51/51 passing ✓
- **Breaking**: Form schema IDs (migration required for dependent code)
- **Performance**: Search debounced, no unnecessary API calls
- **UX**: Inline dropdown, auto-create indication, loading states
- **Code Quality**: Type-safe, well-logged, properly tested

### CODE REVIEW

File: frontend/src/features/invoices/InvoiceForm.tsx [hunk 1]
```diff
@@ -218,35 +231,93 @@ export function InvoiceForm() {
...
+                              setCustomerSearchResults(results);
+                              console.log("✅ Search results:", results.length, "items");
+                            } catch (error) {
+                              console.error("❌ Search error:", error);
+                            } finally {
+                              setIsSearchingCustomers(false);
+                            }
+                          }, 300); // 300ms debounce
+
+                          return () => clearTimeout(timer);
```
[the remark this answers is not in the document]
Comment (wontknow) [reply frontend/src/features/invoices/InvoiceForm.tsx]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/wontknow/Billino/pull/87#discussion_r2510320358)
